src/mscorlib/src/System/Collections/Generic/EqualityComparer.cs

```diff
@@ -36,57 +36,37 @@ private static EqualityComparer<T> CreateComparer()
+                // Instead, it relies on EqualityComparer<T?>.Default to specialize for nullables and do the lifted comparisons if T implements IEquatable.
+                if (type.IsValueType)
+                {
+                    // Nullable/enum equality comparers are less common than IEquatable equality comparers.
+                    // Their creation logic is put in a separate, non-generic class to minimize the cost of generic instantiations of this method
+                    // for the more-common codepath.
+                    result = RareComparers.TryCreateNullableEqualityComparer(type);
                 }
             }
-            // See the METHOD__JIT_HELPERS__UNSAFE_ENUM_CAST and METHOD__JIT_HELPERS__UNSAFE_ENUM_CAST_LONG cases in getILIntrinsicImplementation
-            else if (t.IsEnum) {
-                TypeCode underlyingTypeCode = Type.GetTypeCode(Enum.GetUnderlyingType(t));
-
-                // Depending on the enum type, we need to special case the comparers so that we avoid boxing
-                // Note: We have different comparers for Short and SByte because for those types we need to make sure we call GetHashCode on the actual underlying type as the 
-                // implementation of GetHashCode is more complex than for the other types.
-                switch (underlyingTypeCode) {
-                    case TypeCode.Int16: // short
-                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(ShortEnumEqualityComparer<short>), t);
-                        break;
-                    case TypeCode.SByte:
-                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(SByteEnumEqualityComparer<sbyte>), t);
-                        break;
-                    case TypeCode.Int32:
-                    case TypeCode.UInt32:
-                    case TypeCode.Byte:
-                    case TypeCode.UInt16: //ushort
-                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(EnumEqualityComparer<int>), t);
-                        break;
-                    case TypeCode.Int64:
-                    case TypeCode.UInt64:
-                        result = RuntimeTypeHandle.CreateInstanceForAnotherGenericParameter((RuntimeType)typeof(LongEnumEqualityComparer<long>), t);
-                        break;
-                }
+            else if (type.IsEnum)
+            {
+                // The equality comparer for enums is specialized to avoid boxing.
+                result = RareComparers.TryCreateEnumEqualityComparer(type);
             }
             
-            return result != null ?
-                (EqualityComparer<T>)result :
-                new ObjectEqualityComparer<T>(); // Fallback to ObjectEqualityComparer, which uses boxing
+            return result != null ? (EqualityComparer<T>)result : new ObjectEqualityComparer<T>();
```

Comment:
It is unfortunate that we will create ObjectEqualityComparer type because of it is referenced, even though it will be unused in many cases. We should create the ObjectEqualityComparer using CreateInstanceForAnotherGenericParameter like all other comparers.